File: android/src/main/kotlin/com/kurban/xue_hua_speaker_earpiece_toggle/XueHuaSpeakerEarpieceTogglePlugin.kt
package com.kurban.xue_hua_speaker_earpiece_toggle

import io.flutter.embedding.engine.plugins.FlutterPlugin
import io.flutter.plugin.common.EventChannel
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import io.flutter.plugin.common.MethodChannel.MethodCallHandler
import io.flutter.plugin.common.MethodChannel.Result

/** XueHuaSpeakerEarpieceTogglePlugin */
class XueHuaSpeakerEarpieceTogglePlugin :
    FlutterPlugin,
    MethodCallHandler {
    private lateinit var channel: MethodChannel
    private lateinit var eventChannel: EventChannel
    private lateinit var controller: AudioRouteController
    private lateinit var routeChangeNotifier: RouteChangeNotifier

    override fun onAttachedToEngine(flutterPluginBinding: FlutterPlugin.FlutterPluginBinding) {
        channel = MethodChannel(
            flutterPluginBinding.binaryMessenger,
            "xue_hua_speaker_earpiece_toggle"
        )
        eventChannel = EventChannel(
            flutterPluginBinding.binaryMessenger,
            "xue_hua_speaker_earpiece_toggle/events"
        )
        channel.setMethodCallHandler(this)
        controller = AudioRouteController(flutterPluginBinding.applicationContext)
        routeChangeNotifier = RouteChangeNotifier(
            flutterPluginBinding.applicationContext,
            controller
        )
        eventChannel.setStreamHandler(routeChangeNotifier)
    }

    override fun onMethodCall(
        call: MethodCall,
        result: Result
    ) {
        when (call.method) {
            "getRoute" -> result.success(controller.getRoute())
            "setRoute" -> handleSetRoute(call, result)
            else -> result.notImplemented()
        }
    }

    override fun onDetachedFromEngine(binding: FlutterPlugin.FlutterPluginBinding) {
        eventChannel.setStreamHandler(null)
        controller.restoreSessionIfNeeded()
        channel.setMethodCallHandler(null)
    }

    private fun handleSetRoute(call: MethodCall, result: Result) {
        val route = when (val rawRoute = call.argument<Any?>("route")) {
            is String -> rawRoute
            null -> {
                result.error(
                    "INVALID_ARGUMENT",
                    "Missing route argument.",
                    null
                )
                return
            }
            else -> {
                result.error(
                    "INVALID_ARGUMENT",
                    "Route argument must be a string.",
                    null
                )
                return
            }
        }

        if (route !in AudioRouteNames.SWITCHABLE_ROUTES) {
            result.error(
                "INVALID_ROUTE",
                "Unknown audio route: $route",
                null
            )
            return
        }

        try {
            result.success(controller.setRoute(route).toMap())
        } catch (exception: IllegalArgumentException) {
            result.error(
                "INVALID_ROUTE",
                exception.message,
                null
            )
        } catch (exception: Exception) {
            result.error(
                "AUDIO_ROUTE_ERROR",
                exception.message,
                null
            )
        }
    }
}

File: ios/xue_hua_speaker_earpiece_toggle/Sources/xue_hua_speaker_earpiece_toggle/XueHuaSpeakerEarpieceTogglePlugin.swift
import Flutter
import UIKit

public class XueHuaSpeakerEarpieceTogglePlugin: NSObject, FlutterPlugin {
  private let controller = AudioRouteController()
  private lazy var routeChangeStreamHandler = RouteChangeStreamHandler(
    controller: controller
  )

  public static func register(with registrar: FlutterPluginRegistrar) {
    let instance = XueHuaSpeakerEarpieceTogglePlugin()
    let methodChannel = FlutterMethodChannel(
      name: "xue_hua_speaker_earpiece_toggle",
      binaryMessenger: registrar.messenger()
    )
    let eventChannel = FlutterEventChannel(
      name: "xue_hua_speaker_earpiece_toggle/events",
      binaryMessenger: registrar.messenger()
    )
    registrar.addMethodCallDelegate(instance, channel: methodChannel)
    eventChannel.setStreamHandler(instance.routeChangeStreamHandler)
  }

  deinit {
    controller.restoreSessionIfNeeded()
  }

  public func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
    switch call.method {
    case "getRoute":
      result(controller.getRoute())
    case "setRoute":
      guard let arguments = call.arguments as? [String: Any] else {
        result(
          FlutterError(
            code: "INVALID_ARGUMENT",
            message: "Missing route argument.",
            details: nil
          )
        )
        return
      }

      guard let route = arguments["route"] as? String else {
        result(
          FlutterError(
            code: "INVALID_ARGUMENT",
            message: "Route argument must be a string.",
            details: nil
          )
        )
        return
      }

      do {
        let applyResult = try controller.setRoute(route)
        result(applyResult.dictionary)
      } catch let error as AudioRouteError {
        let flutterError = error.flutterError
        result(
          FlutterError(
            code: flutterError.code,
            message: flutterError.message,
            details: nil
          )
        )
      } catch {
        result(
          FlutterError(
            code: "AUDIO_ROUTE_ERROR",
            message: error.localizedDescription,
            details: nil
          )
        )
      }
    default:
      result(FlutterMethodNotImplemented)
    }
  }
}